ztools/ztls/handshake_client.go

```diff
@@ -32,19 +32,235 @@ type clientHandshakeState struct {
+				c.Extensions[i] = SniExtension{[]string{config.ServerName}, true}
+			}
+			if config.ServerName == "" && len(casted.Domains) > 0 {
+				config.ServerName = casted.Domains[0]
+			}
+		case ALPNExtension:
+			config.NextProtos = casted.Protocols
+		case ExtendedMasterSecretExtension:
+			config.ExtendedMasterSecret = true
+		case SignatureAlgorithmExtension:
+			supportedSKXSignatureAlgorithms = casted.getStructuredAlgorithms()
+			defaultSKXSignatureAlgorithms = casted.getStructuredAlgorithms()
+		case SCTExtension:
+			config.SignedCertificateTimestampExt = true
+		case SupportedCurvesExtension:
+			config.CurvePreferences = casted.Curves
+		case SessionTicketExtension:
+			config.ForceSessionTicketExt = true
+		default:
+			continue
+		}
+	}
+	return config
+}
+
+func (c *ClientFingerprintConfiguration) marshal(config *Config) ([]byte, error) {
+	if err := c.CheckImplementedExtensions(); err != nil {
+		return nil, err
+	}
+	head := make([]byte, 38)
+	head[0] = 1
+	head[4] = uint8(c.HandshakeVersion >> 8)
+	head[5] = uint8(c.HandshakeVersion)
+	if len(c.ClientRandom) == 32 {
+		copy(head[6:38], c.ClientRandom[0:32])
+	} else {
+		_, err := io.ReadFull(config.rand(), head[6:38])
+		if err != nil {
+			return nil, errors.New("tls: short read from Rand: " + err.Error())
+		}
+	}
+
+	sessionID := make([]byte, len(c.SessionID)+1)
+	sessionID[0] = uint8(len(c.SessionID))
+	if len(c.SessionID) > 0 {
+		copy(sessionID[1:], c.SessionID)
+	}
+
+	ciphers := make([]byte, 2+2*len(c.CipherSuites))
+	ciphers[0] = uint8(len(c.CipherSuites) >> 7)
+	ciphers[1] = uint8(len(c.CipherSuites) << 1)
+	for i, suite := range c.CipherSuites {
+		if !config.ForceSuites {
+			found := false
+			for _, impl := range implementedCipherSuites {
+				if impl.id == suite {
+					found = true
+				}
+			}
+			if !found {
+				return nil, errors.New(fmt.Sprintf("tls: unimplemented cipher suite %d", suite))
+			}
+		}
+
+		ciphers[2+i*2] = uint8(suite >> 8)
+		ciphers[3+i*2] = uint8(suite)
+	}
+
+	compressions := make([]byte, len(c.CompressionMethods)+1)
+	compressions[0] = uint8(len(c.CompressionMethods))
+	if len(c.CompressionMethods) > 0 {
+		copy(compressions[1:], c.CompressionMethods)
+		if c.CompressionMethods[0] != 0 {
+			return nil, errors.New(fmt.Sprintf("tls: unimplemented compression method %d", c.CompressionMethods[0]))
+		} else if len(c.CompressionMethods) > 1 {
+			return nil, errors.New(fmt.Sprintf("tls: unimplemented compression method %d", c.CompressionMethods[1]))
+		}
+	} else {
+		return nil, errors.New("tls: no compression method")
+	}
+
+	var extensions []byte
+	for _, ext := range c.Extensions {
+		extensions = append(extensions, ext.Marshal()...)
+	}
+	if len(extensions) > 0 {
+		length := make([]byte, 2)
+		length[0] = uint8(len(extensions) >> 8)
+		length[1] = uint8(len(extensions))
+		extensions = append(length, extensions...)
+	}
+	hello := append(head, append(sessionID, append(ciphers, append(compressions, extensions...)...)...)...)
```

Comment:
Not really. Although, preallocating a slice and then appending in a loop of a [][]byte might read a little easier and be faster? Alternatively, you could try to preallocate everything before you serialize, and then just slice off the relevant chunks as you go.